tests/unit/test_shadow_account_policy.py

```diff
@@ -0,0 +1,486 @@
+def test__password_complexity(ruleset, password, expected, set_fips_available):
+    """ Make sure password deficiencies are properly identified """
+    assert check_password_complexity(ruleset, password) == expected
+
+
+@pytest.mark.parametrize('sec', (
+    {
+        'min_password_age': 10,
+        'max_password_age': None,
+    },
+    {
+        'min_password_age': None,
+        'max_password_age': 60,
+    },
+    {
+        'min_password_age': None,
+        'max_password_age': None,
+    },
+))
+def test__account_policy_shadow_file(sec, set_fips_available):
+    """ Verify that system security account settings properly modify shadow file """
+    with Client() as c:
+        c.call('system.security.update', sec, job=True)
+
+    with create_user('test_user_policy') as u:
+        shadow = get_shadow_entry(u['username'])
+        for key, value in sec.items():
+            assert shadow[key] == value
+
+
+# NOTE: this can be expanded when middleware has more comprensive PAM plumbing
+# to allow password resets / expose shadow warning and inactive fields
+
+
+@pytest.mark.parametrize('sec,error', (
+    ({
+        'min_password_age': 25,
+        'max_password_age': 10,
+    }, 'Minimum password age must be lower than the maximum password age'),
+))
+def test__account_policy_validation(sec, error, set_fips_available):
+    with Client() as c:
+        with pytest.raises(Exception, match=error):
+            c.call('system.security.update', sec, job=True)
+
+
+def test__password_history(password_history):
+    """ Verify that password history is properly applied """
+    with create_user('history_user', random_password=False, password='Canary') as u:
+        with Client() as c:
+            with pytest.raises(Exception, match='requires a password that does not match'):
+                c.call('user.update', u['id'], {'password': 'Canary'})
+
+            for i in range(0, TEST_HISTORY):
+                c.call('user.update', u['id'], {'password': f'Canary{i}'})
+
+            # We've cycled through history and should be OK again
+            c.call('user.update', u['id'], {'password': 'Canary'})
+
+
+def test__password_expired(old_admin):
+    """ Verify that expired password gives EXPIRED response type """
+    with Client('ws://127.0.0.1/api/current') as c:
+        resp = c.call('auth.login_ex', {
+            'mechanism': 'PASSWORD_PLAIN',
+            'username': old_admin['username'],
+            'password': old_admin['password']
+        })
+        assert resp['response_type'] == 'SUCCESS'
+
+    with security_config(max_password_age=60):
+        with Client() as c:
+            user = c.call('user.get_instance', old_admin['id'])
+            assert user['password_change_required']
+
+            # expired accounts should generate an alert
+            alerts = c.call('alert.run_source', 'SecurityLocalUserAccountExpiration')
+            assert any([alert['klass'] == 'LocalAccountExpired' for alert in alerts]), str(alerts)
```

Comment:
Might be an idea to also validate that no such alerts exist earlier in this test.